File: src/pi/modules/control_tasks/telemetry_control.py
import json, enum, time
from heapq import heappop
from modules.mcl.flag import Flag
from modules.lib.errors import Error
from modules.lib.helpers import enqueue
from modules.mcl.registry import Registry
from modules.lib.packet import Packet, Log, LogPriority
from modules.lib.enums import SensorType, SensorLocation, ValveLocation, ActuationType, ValveType, ValvePriority, Stage

class TelemetryControl(): 
    def __init__(self, registry: Registry, flag: Flag):
        self.registry = registry
        self.flag = flag
        # TODO: Add an "undo_soft_abort" function
        self.funcs = {
            "heartbeat": self.heartbeat,
            "soft_abort": self.soft_abort,
            "solenoid_actuate": self.solenoid_actuate,
            "sensor_request": self.sensor_request,
            "valve_request": self.valve_request,
            "progress": self.progress,
            "test": self.test,
        }
        self.arguments = {
            "heartbeat": (),
            "soft_abort": (),
            "solenoid_actuate": (("valve_location", ValveLocation), ("actuation_type", ActuationType), ("priority", ValvePriority)),
            "sensor_request": (("sensor_type", SensorType), ("sensor_location", SensorLocation)),
            "valve_request": (("valve_type", ValveType), ("valve_location", ValveLocation)),
            "progress": (),
            "test": (("response", str),)
        }
    

    def begin(self, config: dict):
        self.config = config


    def execute(self) -> Error:
        #TODO: Check if resetting telemetry works
        _, status, timestamp = self.registry.get(("telemetry", "status"))
        if not status:
            self.flag.put(("telemetry", "reset"), True)
            return None

        self.flag.put(("telemetry", "reset"), False)
        _, telem_queue, _ = self.registry.get(("telemetry", "ingest_queue"))
        while telem_queue:
            packet = heappop(telem_queue)
            #TODO: Figure out if the command from a log is outdated
            for log in packet.logs:
                print(log)
                err = self.ingest(log)
            # Clear the ingest queue (because we just ingested everything)
            self.registry.put(("telemetry", "ingest_queue"), [])


    def ingest(self, log: Log):
        header = log.header
        if header in self.funcs:
            func = self.funcs[header]
            args = []
            assert(header in self.arguments)
            # Loop through each of the arguments
            for arg_name, arg_type in self.arguments[header]:
                # If you can't find the argument in the 'message' section, throw an exception.
                if arg_name not in log.message:
                    print("Invalid argument", arg_name)
                    log = Log(header="response", message={"header": "Missing argument", "Argument": arg_name})
                    enqueue(self.flag, log, LogPriority.CRIT)
                    return Error.INVALID_ARGUMENT_ERROR
                # If its an enum, then make it an enum
                if issubclass(arg_type, enum.Enum):
                    if log.message[arg_name] not in [x.value for x in arg_type]:
                        print("Invalid argument", arg_name, arg_type)
                        log = Log(header="response", message={"header": "Invalid argument type", "Argument": arg_name, "Received argument type": str(type(log.message[arg_name])), "Expected argument type (enum)": str(arg_type)})
                        enqueue(self.flag, log, LogPriority.CRIT)
                        return Error.INVALID_ARGUMENT_ERROR
                # Otherwise, take the raw value
                elif not isinstance(log.message[arg_name], arg_type):
                    print("Invalid argument", arg_name, arg_type)
                    log = Log(header="response", message={"header": "Invalid argument type", "Argument": arg_name, "Received argument type": str(type(log.message[arg_name])), "Expected argument type": str(arg_type)})
                    enqueue(self.flag, log, LogPriority.CRIT)
                    return Error.INVALID_ARGUMENT_ERROR
                args.append(arg_type(log.message[arg_name]))
            # Run the command w/ the given arguments
            print("Running command (via telemetry):", header, *args)
            func(*args)
            return Error.NONE
        else:
            print("Invalid header")
            log = Log(header="response", message={"header": "Invalid telemetry header", "Telemetry header": header})
            enqueue(self.flag, log, LogPriority.CRIT)
            return Error.INVALID_HEADER_ERROR


    def heartbeat(self):
        enqueue(self.flag, Log(header="heartbeat", message={"response": "OK", "mode": "Soft abort" if self.registry.get(("general", "soft_abort"))[1] else "Normal"}), LogPriority.INFO)
        print("mode", self.registry.get(("general", "soft_abort"))[1])


    def soft_abort(self):
        self.registry.put(("general", "soft_abort"), True)
        log = Log(header="response", message={"header": "Soft abort", "Status": "Success", "Description": "Rocket is undergoing soft abort"})
        enqueue(self.flag, log, LogPriority.CRIT)
        enqueue(self.flag, Log(header="mode", message={"mode": "Soft abort"}), LogPriority.CRIT)


    def solenoid_actuate(self, valve_location: ValveLocation, actuation_type: ActuationType, priority: int) -> Error:
        err, current_priority, timestamp = self.registry.get(("valve_actuation", "actuation_priority", ValveType.SOLENOID, valve_location), allow_error=True)
        if err != Error.NONE:
            # Send message back to gs saying it was an invalid message
            log = Log(header="response", message={"header": "Valve actuation", "Status": "Failure", "Description": "Unable to find actuatable solenoid", "Provided valve location": valve_location})
            enqueue(self.flag, log, LogPriority.CRIT)
            return Error.REQUEST_ERROR

        if priority > current_priority:
            # Send message back to gs saying that the request was made w/ too little priority
            log = Log(header="response", message={"header": "Valve actuation", "Status": "Failure", "Description": "Too little priority to actuate solenoid", "Valve location": valve_location, "Actuation type": actuation_type, "Priority": priority})
            enqueue(self.flag, log, LogPriority.CRIT)
            return Error.PRIORITY_ERROR

        print("Actuating solenoid at {} with actuation type {}".format(valve_location, actuation_type))

        self.flag.put(("solenoid", "actuation_type", valve_location), actuation_type)
        self.flag.put(("solenoid", "actuation_priority", valve_location), priority)

        log = Log(header="response", message={"header": "Valve actuation", "Status": "Success", "Description": "Successfully actuated solenoid", "Valve location": valve_location, "Actuation type": actuation_type, "Priority": priority})
        enqueue(self.flag, log, LogPriority.CRIT)


    def sensor_request(self, sensor_type: SensorType, sensor_location: SensorLocation) -> Error:
        err, value, timestamp = self.registry.get(("sensor_measured", sensor_type, sensor_location), allow_error=True)
        if err != Error.NONE:
            # Send message back to gs saying it was an invalid message            
            log = Log(header="response", message={"header": "Sensor data", "Status": "Failure", "Description": "Unable to find sensor", "Type": sensor_type, "Location": sensor_location})
            enqueue(self.flag, log, LogPriority.CRIT)
            return Error.REQUEST_ERROR

        _, kalman_value, _ = self.registry.get(("sensor_normalized", sensor_type, sensor_location), allow_error=True)
        _, status, _ = self.registry.get(("sensor_status", sensor_type, sensor_location))
        log = Log(header="response", message={"header": "Sensor data", "Status": "Success", "Sensor type": sensor_type, "Sensor location": sensor_location, "Measured value": value, "Normalized value": kalman_value, "Sensor status": status, "Last updated": timestamp})
        enqueue(self.flag, log, LogPriority.INFO)


    def valve_request(self, valve_type: ValveType, valve_location: ValveLocation) -> Error: 
        err, value, timestamp = self.registry.get(("valve", valve_type, valve_location), allow_error=True)
        if err != Error.NONE:
            # Send message back to gs saying it was an invalid message            
            log = Log(header="response", message={"header": "Valve data request", "Status": "Failure", "Description": "Unable to find valve", "Valve type": valve_type, "Valve location": valve_location})
            enqueue(self.flag, log, LogPriority.CRIT)
            return Error.REQUEST_ERROR

        _, actuation_type, timestamp = self.registry.get(("valve_actuation", "actuation_type", valve_type, valve_location))
        _, actuation_priority, _ = self.registry.get(("valve_actuation", "actuation_priority", valve_type, valve_location))
        log = Log(header="response", message={"header": "Valve data", "Status": "Success", "Type": valve_type, "Location": valve_location, "State": value, "Actuation type": actuation_type, "Actuation priority": actuation_priority, "Last actuated": timestamp})
        enqueue(self.flag, log, LogPriority.INFO)


    def progress(self):
        self.flag.put(("general", "progress"), True)


    def test(self, msg: str):
        print("\ntest recieved:", msg)



"""
    def undo_soft_abort(self):
        # GO THROUGH EACH SENSOR AND CHECK STATUS
        sensor_list = self.registry.values["sensor_status"]
        critical_sensors = []
        for s_type in sensor_list:
            for s_loc in s_type:
                if sensor_list[s_type][s_loc] == LogPriority.CRIT:
                    critical_sensors.append([s_type, s_loc])
        
        # SEND FAILURE OR SUCCESS LOG DEPENDING ON CRITICAL VALUES FOUND/NOT FOUND
        if len(critical_sensors) > 0:
            log = Log(header="response", message={"header": "Soft abort cancellation", "Status": "Failure", "Description": "Soft abort cancellation unsuccessful - following sensors are still not safe: {}".format(critical_sensors)})
            enqueue(self.flag, log, LogPriority.CRIT)
        else:
            self.registry.put(("general", "soft_abort"), None)
            log = Log(header="response", message={"header": "Soft abort cancellation", "Status": "Success", "Description": "Soft abort successfully cancelled"})
            enqueue(self.flag, log, LogPriority.CRIT)
            enqueue(self.flag, Log(header="mode", message={"mode": "Normal"}), LogPriority.CRIT)
"""

File: src/modules/control_tasks/TelemetryControl.cpp
//
// Created by adiv413 on 4/24/2020.
//

#include <flight/modules/control_tasks/TelemetryControl.hpp>
#include <flight/modules/lib/Util.hpp>
#include <Logger/logger_util.h>
#include <queue>

using nlohmann::json;

//TODO: add custom packet enqueuing interface to gs???

TelemetryControl::TelemetryControl() {
    global_flag.log_info("response", {
        {"header", "info"},
        {"Description", "Telemetry Control started"}
    });
}

void TelemetryControl::begin() {
    log("Telemetry control: Beginning");
    telemetry.connect();
    make_functions();
}

// Store list of all commands that GS can send as functions, add the function pointers to the map and call when necessary
void TelemetryControl::make_functions() {
    log("Telemetry: Making Functions");
    this->functions.emplace("heartbeat", &TelemetryControl::heartbeat);
    this->functions.emplace("soft_abort", &TelemetryControl::soft_abort);
    this->functions.emplace("solenoid_actuate", &TelemetryControl::solenoid_actuate);
    this->functions.emplace("sensor_request", &TelemetryControl::sensor_request);
    this->functions.emplace("valve_request", &TelemetryControl::valve_request);
    this->functions.emplace("progress", &TelemetryControl::progress);
    this->functions.emplace("test", &TelemetryControl::test);
}

void TelemetryControl::execute() {
    log("Telemetry control: Executing");
    if (!global_registry.telemetry.status) {
        global_flag.telemetry.reset = true;
    } else {
        global_flag.telemetry.reset = false;
        auto &ingest_queue = global_registry.telemetry.ingest_queue;
        while (!ingest_queue.empty()) {
            Packet packet = ingest_queue.top();

            //TODO: figure out if log command is outdated
            for(const Log& log_ : packet.getLogs()) {
                json j;
                to_json(j, log_);
                log(j.dump());
                ingest(log_);
            }
        }
    }
}
void TelemetryControl::ingest(const Log& log) {
    string header = log.getHeader();
    json params = log.getMessage();

    // Make sure the function exists
    if (this->functions.find(header) == this->functions.end()) {
        throw INVALID_HEADER_ERROR();
    }

    auto function = this->functions.at(header);
    vector<string> argument_order = arguments.at(header);

    // TODO: change the packet format from gs to make it strings instead of enums

    if (argument_order.size() != params.size()) {
        throw PACKET_ARGUMENT_ERROR();
    }

    vector<string> param_values;

    try {
        for (const string& argument_name : argument_order) {
            param_values.push_back(params.at(argument_name).get<string>());
        }
    } catch (...) {
        global_flag.log_warning("invalid_argument", {{"message", "Invalid function arguments"}});
    }

    (this->*function)(param_values); // call function which maps to the GS command sent w/ all params necessary
}
void TelemetryControl::heartbeat(const vector<string>& args) {
    global_flag.log_info("heartbeat", {
        {"header", "heartbeat"},
        {"response", "OK"}
    });
}

void TelemetryControl::soft_abort(const vector<string>& args) {
    global_registry.general.soft_abort = true;
    global_flag.log_critical("response", {
        {"header", "Soft Abort"},
        {"Status", "Success"},
        {"Description", "Rocket is undergoing soft abort"}
    });
    global_flag.log_critical("mode", {
        {"header", "Soft Abort"},
        {"mode", "Soft Abort"}
    });
}
void TelemetryControl::solenoid_actuate(const vector<string>& args) {
    if (!global_registry.valve_exists("solenoid", args[0])) {
        global_flag.log_critical("Valve actuation",{
            {"header", "Valve actuation"},
            {"Status", "Failure"},
            {"Description", "Unable to find actuatable solenoid"},
            {"Valve location", args[0]}
        });
        throw INVALID_SOLENOID_ERROR();
    }

    int current_priority = int(global_registry.valves["solenoid"][args[0]].actuation_priority);

    if (int(valve_priority_map[args[2]]) < current_priority) {
        global_flag.log_critical("Valve actuation", {
            {"header", "Valve actuation"},
            {"Status", "Failure"},
            {"Description", "Priority too low to actuate"},
            {"Valve location", args[0]},
            {"Actuation type", args[1]},
            {"Priority", args[2]}
        });
        throw INSUFFICIENT_PRIORITY_SOLENOID_ERROR();
    }

    log("Actuating solenoid at " + args[0] + " with actuation type " + args[1]);

    try {
        //TODO: make sure gs packets have the upper case version of the enum as the value for the actuation type
        FlagValveInfo &valve_flag = global_flag.valves["solenoid"][args[0]];
        valve_flag.actuation_type = actuation_type_map[args[1]];
        valve_flag.actuation_priority = valve_priority_map[args[2]];
    } catch(...) {
        global_flag.log_critical("Valve actuation", {
            {"header", "Valve actuation"},
            {"Status", "Failure"},
            {"Description", "Wrong packet message"},
            {"Valve location", args[0]},
            {"Actuation type", args[1]},
            {"Priority", args[2]}
        });
        throw INVALID_PACKET_MESSAGE_ERROR();
    }

    global_flag.log_info("Valve actuation", {
        {"header", "Valve actuation"},
        {"Status", "Success"},
        {"Description", "Successfully activated solenoid"}
    });
}

void TelemetryControl::sensor_request(const vector<string>& args) {
    double value;
    double kalman_value;
    string sensor_status_str;
    string sensor_type = args[0];
    string sensor_loc = args[1];

    if (!global_registry.sensor_exists(sensor_type, sensor_loc)) {
        global_flag.log_critical("response", {
            {"header", "Sensor data"},
            {"Status", "Failure"},
            {"Description", "Unable to find sensor"},
            {"Sensor type", args[0]},
            {"Sensor location", args[1]}
        });

        throw INVALID_SENSOR_LOCATION_ERROR();
    }

    auto sensor = global_registry.sensors[sensor_type][sensor_loc];
    value = sensor.measured_value;
    kalman_value = sensor.normalized_value;
    sensor_status_str = sensor_status_map[sensor.status];

    global_flag.log_critical("response", {
        {"header", "Sensor data request"},
        {"Status", "Success"},
        {"Sensor type", args[0]},
        {"Sensor location", args[1]},
        {"Sensor status", sensor_status_str},
        {"Measured value", value},
        {"Normalized value", kalman_value},
        {"Last updated", std::chrono::system_clock::now().time_since_epoch().count()}
    });
}
void TelemetryControl::valve_request(const vector<string>& args) {
    string state;
    string actuation_type;
    string actuation_priority;
    string valve_type = args[0];
    string valve_loc = args[1];

    if (!global_registry.valve_exists(valve_type, valve_loc)) {
        global_flag.log_critical("response", {
            {"header", "Valve data request"},
            {"Status", "Failure"},
            {"Description", "Unable to find valve"},
            {"Valve type", valve_type},
            {"Valve location", valve_loc}
        });
        throw INVALID_VALVE_LOCATION_ERROR();
    }

    auto valve_registry = global_registry.valves[valve_type][valve_loc];

    actuation_type = actuation_type_inverse_map.at(valve_registry.actuation_type);
    actuation_priority = valve_priority_inverse_map.at(valve_registry.actuation_priority);

    global_flag.log_critical("response", {
        {"header", "Valve data request"},
        {"Status", "Success"},
        {"Actuation type", actuation_type},
        {"Actuation priority", actuation_priority},
        {"Valve type", valve_type},
        {"Valve location", valve_loc},
        {"Last actuated", std::chrono::system_clock::now().time_since_epoch().count()}
    });
}
void TelemetryControl::progress(const vector<string>& args) {
    global_flag.general.progress = true;
}
void TelemetryControl::test(const vector<string>& args) {
    log("Test received: " + args[0]);
}